From 2799b4535328ec154a11dc510959746f73b30212 Mon Sep 17 00:00:00 2001
Message-Id: <2799b4535328ec154a11dc510959746f73b30212.1413837857.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: Vidya Srinivas <vidya.srinivas@intel.com>
Date: Thu, 10 Jul 2014 16:04:26 +0530
Subject: [PATCH 154/312] FOR_UPSTREAM: [VPG]: drm/i915: Fix sideband write
 warnings

The following changes have been ported:
	Original Change-ID:Ia448ad4d5841101d39726fcb8f234b22c5c8df4e
	DPIO Sideband writes create warnings due to incorrect
	mutex lock. Patch fixes the issue.

Issue: GMIN-2364
Change-Id: Ia7a77500fd66bff6c08ffe310230057092adbcfc
Signed-off-by: Shobhit Kumar <shobhit.kumar@intel.com>
Signed-off-by: Vidya Srinivas <vidya.srinivas@intel.com>
---
 drivers/gpu/drm/i915/i915_dma.c       |    1 +
 drivers/gpu/drm/i915/i915_drv.h       |    1 +
 drivers/gpu/drm/i915/intel_sideband.c |    5 ++++-
 3 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_dma.c b/drivers/gpu/drm/i915/i915_dma.c
index 0c01896..b293deca 100644
--- a/drivers/gpu/drm/i915/i915_dma.c
+++ b/drivers/gpu/drm/i915/i915_dma.c
@@ -1607,6 +1607,7 @@ int i915_driver_load(struct drm_device *dev, unsigned long flags)
 	spin_lock_init(&dev_priv->mm.object_stat_lock);
 	spin_lock_init(&dev_priv->mmio_flip_lock);
 	mutex_init(&dev_priv->dpio_lock);
+	mutex_init(&dev_priv->new_dpio_lock);
 	mutex_init(&dev_priv->modeset_restore_lock);
 	mutex_init(&dev_priv->dpst.ioctl_lock);
 
diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 7a3534b..c3b9720 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -1472,6 +1472,7 @@ struct drm_i915_private {
 
 	/* DPIO indirect register protection */
 	struct mutex dpio_lock;
+	struct mutex new_dpio_lock;
 
 	/** Cached value of IMR to avoid reads in updating the bitfield */
 	union {
diff --git a/drivers/gpu/drm/i915/intel_sideband.c b/drivers/gpu/drm/i915/intel_sideband.c
index fd0cf1d..8f4de37 100644
--- a/drivers/gpu/drm/i915/intel_sideband.c
+++ b/drivers/gpu/drm/i915/intel_sideband.c
@@ -56,12 +56,13 @@ static int vlv_sideband_rw(struct drm_i915_private *dev_priv, u32 devfn,
 		(port << IOSF_PORT_SHIFT) | (be << IOSF_BYTE_ENABLES_SHIFT) |
 		(bar << IOSF_BAR_SHIFT);
 
-	WARN_ON(!mutex_is_locked(&dev_priv->dpio_lock));
+	mutex_lock(&dev_priv->new_dpio_lock);
 
 	if (wait_for_atomic((I915_READ(VLV_IOSF_DOORBELL_REQ)
 					& IOSF_SB_BUSY) == 0, 5)) {
 		DRM_DEBUG_DRIVER("IOSF sideband idle wait (%s) timed out\n",
 				 is_read ? "read" : "write");
+		mutex_unlock(&dev_priv->new_dpio_lock);
 		return -EAGAIN;
 	}
 
@@ -74,12 +75,14 @@ static int vlv_sideband_rw(struct drm_i915_private *dev_priv, u32 devfn,
 					& IOSF_SB_BUSY) == 0, 5)) {
 		DRM_DEBUG_DRIVER("IOSF sideband finish wait (%s) timed out\n",
 				 is_read ? "read" : "write");
+		mutex_unlock(&dev_priv->new_dpio_lock);
 		return -ETIMEDOUT;
 	}
 
 	if (is_read)
 		*val = I915_READ(VLV_IOSF_DATA);
 	I915_WRITE(VLV_IOSF_DATA, 0);
+	mutex_unlock(&dev_priv->new_dpio_lock);
 
 	return 0;
 }
-- 
1.7.9.5

